### PR TITLE
Ease development via README, CI, and lockfile

### PR DESCRIPTION
**Pull request by samestep**
Currently it's a bit unclear how to work on this project, so this PR attempts to fix that in the following ways:

- Specify dependencies (both for `ghstack` itself and for development of `ghstack`) in `pyproject.toml` (which replaces `setup.py` and also becomes the source of truth for `ghstack.__version__`)
- Expand the "Developer notes" section of the README to include instructions on how to install those dependencies, run the tests, and publish to PyPI
- Add a GitHub Actions workflow that essentially follows those same steps (to install dependencies and test, but not publish)
- Add a `poetry.lock`file to make it more likely that build/test success in CI implies build/test success on a local development machine

To get CI to pass, I needed to fix a couple things that were broken:

- `run_tests.sh` referred to `flake8-3`, which I don't seem to get when installing Flake8 from PyPI
- `ghstack/land.py` had a minor `mypy` error
- `test_ghstack.py` had a few tests that failed with recent versions of Git, due to changes in the behavior of `git log --graph` introduced sometime between Git 2.24 and Git 2.30

Currently this PR involves switching from setuptools to Poetry, but that's just because Poetry makes it much easier to maintain a lockfile. However, one reason (besides simply avoiding friction) to stick with setuptools might be that [Poetry doesn't currently support `scripts`](https://github.com/python-poetry/poetry/issues/241), so this PR includes an implicit revert of 55e61892396d0471a6bb0e705282f5c7a0cbbdcf (since `[tool.poetry.scripts]` corresponds to `entry_points` rather than `scripts`).

> This is not supported by pip 10, which a lot of people have installed (because it's the conda default)

Is this comment about pip 10 still true, or would the `entry_points` generated by Poetry work fine nowadays? Also, more generally, here are all the differences between the two generated sdists:

- `old`: https://pypi.org/project/ghstack/0.3.3/#files
- `new`: https://test.pypi.org/project/ghstack/0.3.3/#files
- sdist diff: https://pastebin.com/XpjjHiTy

@ezyang tl;dr if you'd rather not switch the build system for this repo, just let me know and I'll update this PR to try to accomplish the same goals using setuptools. In any case, I would like to merge some variation of this PR, including at least some better README instructions and a GitHub Actions workflow.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,8 +59,31 @@ pull request and pushes commits onto three branches:
 
 ## Developer notes
 
+This project uses [Poetry](https://python-poetry.org/docs/#installation), so
+after you've installed Poetry itself, run this command in your clone of this
+repo to install all the dependencies you need for working on `ghstack`:
+```
+poetry install
```
Comment (ezyang) [inline README.md:66]: Should probably mention how to actually, ya know, run ghstack after you do poetry (unlike python setup.py develop it won't dunk a ghstack binary in your bin)

**Comment by ezyang** [comment]
> Is this comment about pip 10 still true, or would the entry_points generated by Poetry work fine nowadays? A

Yeah, this would probably be a dealbreaker for poetry if it was still true. However, the commit is from 2019, and I quickly looked over all of my Python environments and I didn't see any where pip 10 was a thing, so probably we can just YOLO this and fix it later if someone actually complains. It's a bit hard to understand the linked poetry issue as everyone is confused about whether or not it's for making `poetry test` working or the distributed wheel (which *kinda* actually seems like it's landed?)

The one last blocker is when I check out your PR and try to run some tests, most of them are failing like below. Not sure what's up with that.

```
======================================================================
ERROR: test_update_fields_preserves_commit_message (__main__.TestGh)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ezyang/Dev/ghstack/test_ghstack.py", line 82, in setUp
    self.substituteRev(GitCommitHash("HEAD"), SubstituteRev("rINI0"))
  File "/Users/ezyang/Dev/ghstack/test_ghstack.py", line 94, in substituteRev
    h = GitCommitHash(self.sh.git("rev-parse", "--short", rev))
  File "/Users/ezyang/Dev/ghstack/ghstack/shell.py", line 272, in git
    return self._maybe_rstrip(self.sh(*(("git",) + args), **kwargs))
  File "/Users/ezyang/Dev/ghstack/ghstack/shell.py", line 211, in sh
    raise RuntimeError(
RuntimeError: git rev-parse --short HEAD failed with exit code 128

----------------------------------------------------------------------
```

File: README.md (hunk 1 above)
Comment (samestep) [reply README.md:66]: Good point! Fixed.

**Review by ezyang — APPROVED** [review]
gotta turn on github actions

**Pull request merged**